### PR TITLE
Add user guide: How to identify GPU waste - Create gpu_efficiency.md

### PR DESCRIPTION
**Pull request by philippepicard-stack**

### CODE REVIEW

File: docs/userguides/gpu_efficiency.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+---
+title: "Could my job run faster? How to identify GPU waste"
+description: >-
+  Understand GPU efficiency metrics, diagnose underutilization in your jobs,
+  and apply concrete best practices to improve throughput on the Mila cluster.
+---
+
+# Could my job run faster? How to identify GPU waste
+
+## 1. Why should you care about GPU efficiency?
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: Simplement pour standardiser p/r aux autres guides, les titres ne sont pas numérotés

File: docs/userguides/gpu_efficiency.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  unearth hidden bugs in your data loaders or model architectures. You get
+  useful results faster.
+- **For Mila:** Maximizing efficiency frees up cluster nodes, resulting in
+  shorter queue times and more parallel experiments across the institute.
+
+In other words, efficient compute utilization makes Mila research thrive.
+
+## 2. The basics of compute efficiency: Utilization vs. Occupancy
+
+A common pitfall is relying solely on `nvidia-smi`. The trap is that
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: J'ai l'impression que si les utilisateurs faisaient usage de `nvidia-smi`, on serait déjà bien avancé. Je pense que même les stats de `nvidia-smi` montre une sous-utilisations des GPUs. Peut-être reformuler un peu _A common pitfall_?

File: docs/userguides/gpu_efficiency.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+`nvidia-smi` reports "100% Utilization" as soon as even one thread is active
+on the GPU — like saying a bus is "used" because there is one passenger inside.
+
+To better measure actual utilization, look at **Streaming Multiprocessor (SM)
+Occupancy**: how many "seats" in the bus are actually filled. More context is
+available in this
+[external post on GPU utilization metrics](https://www.trainy.ai/blog/gpu-utilization-misleading).
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: Si on n'offre pas de façon simple pour les utilisateurs, je doute qu'ils fassent le travail de récolter et analyser les métriques d'utilisation. Est-ce qu'on attend d'avoir le dashboard en ligne et on le link ici avant de publier / mettre cette section en ligne?

File: docs/userguides/gpu_efficiency.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+During a job, you can `srun` into your allocated node and run:
+
+```bash
+# Basic check
+nvidia-smi
+
+# Better: Check SM occupancy in real time (requires dcgmi or nvitop)
+# High power draw (Watts) is usually a good signal of active GPU utilization.
+nvitop
+```
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md:62]: Est-ce qu'on peut avoir un example de à quoi le graph pourrait ressembler?

Comment (satyaog) [reply docs/userguides/gpu_efficiency.md:62]: Je ne suis pas sûr non plus qu'on cherche à avoir des utilisateurs de `nvidia-smi` / `nvitop`? Dans le sens que c'est un indicatif ponctuelle mais demain est-ce que tu te souviendras des données que ces outils te donne maintenant? Est-ce que c'est données sont représentatifs de tout l'entraînement?

C'est pertinent de mentionner ces outils, mais en #2 ça me semble les mettre un peu trop en avant, comme s'ils étaient centrals dans le workflow des étudiants. Ce pourrait être dans un bulle informative du genre "btw, vous pouvez avoir une idée en temps réel de la consommation des GPUs en utilisant ces outils"?

File: docs/userguides/gpu_efficiency.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   just to avoid writing checkpointing code — this tanks your overall queue
+   priority.
+3. **Over-allocating CPU cores:** Do not request excessive CPU cores (e.g., 40
+   CPUs for 1 GPU) unless your preprocessing explicitly requires it. Mila
+   provides CPU-only nodes if needed.
+4. **Scaling GPUs to fix I/O bottlenecks:** Do not add more GPUs if your
+   pipeline is bottlenecked by storage read latency or CPU preprocessing — you
+   will just idle more hardware.
+5. **Underutilizing VRAM:** Do not run with low batch sizes if your VRAM usage
+   is under 20%. Increase the batch size to maximize SM occupancy.
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: il y a une limite sur la taille des batch size qui peuvent être utiliser sans dégrader les performance du modèle entraîné. la solution peut-être préférablement d'utiliser un plus petit GPU et / ou de pack plus d'entraînement de modèle dans une slurm job.

**Comment by satyaog** [comment]
Question générale, est-ce qu'on serait intéressé à standardiser ce guide avec les skills claudes du projet mila docs? Si oui, une fois qu'on approuve cette PR je pourrais exécuter les skills sur ce guide et ouvrir une autre PR par la suite

File: docs/userguides/gpu_efficiency.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+During a job, you can `srun` into your allocated node and run:
+
+```bash
+# Basic check
+nvidia-smi
+
+# Better: Check SM occupancy in real time (requires dcgmi or nvitop)
+# High power draw (Watts) is usually a good signal of active GPU utilization.
+nvitop
```
Comment (obilaniu) [inline docs/userguides/gpu_efficiency.md]: - `dcgmi` est si exceptionnellement difficile à utiliser que je ne le recommanderais pas, faute d'une explication de comment exactement le faire.
- `nvtop`, pas `nvitop`. Et ce n'est malheureusement pas installé sur les noeuds de calcul.

File: docs/userguides/gpu_efficiency.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+# Could my job run faster? How to identify GPU waste
+
+## 1. Why should you care about GPU efficiency?
+
+Optimizing your GPU usage directly accelerates research velocity. As compute
+power at Mila is a shared resource, efficient jobs on the cluster bring a
+two-sided advantage:
+
+- **For you:** Eliminating bottlenecks speeds up your training times and helps
```
Comment (soline-b) [inline docs/userguides/gpu_efficiency.md:16]: Moreover, optimizing resources for the allocation a user requests means that this user is requesting the minimal compute "block" they can, and would probably be able to have access to an allocation earlier

File: docs/userguides/gpu_efficiency.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+1. **Profile before scaling:** Run a test job with a profiler (WandB or
+   TensorBoard) before launching large sweeps to ensure your data loader
+   saturates the GPU.
+2. **Optimize data pipelines:** Set `num_workers > 0` (2–4 per allocated GPU)
+   and enable `pin_memory=True` in your PyTorch `DataLoader` to prevent GPU
+   stalling.
+3. **Implement checkpointing:** Save training states regularly so jobs can
+   automatically resume after preemption or timeouts without losing previous
+   compute hours.
+4. **Right-size resource requests:** Use lower-tier nodes or MIG
```
Comment (soline-b) [inline docs/userguides/gpu_efficiency.md]: Should we explain which ones are "lower-tier nodes"?

File: docs/userguides/gpu_efficiency.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+   (Multi-Instance GPU) slices for small models or debugging instead of
+   allocating full high-end nodes.
+
+### ❌ DON'Ts — Common pitfalls
+
+1. **Hoarding nodes:** Do not keep high-end GPUs (e.g., H100s) allocated on
```
Comment (soline-b) [inline docs/userguides/gpu_efficiency.md:90]: Shoiuld we be more exhaustive than only naming the H100s?

Comment (philippepicard-stack) [reply docs/userguides/gpu_efficiency.md:90]: it's only one example so I believe that's OK

File: docs/userguides/gpu_efficiency.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+| < 5% | Critical waste |
+| ~10% | Poor utilization — the GPU is mostly waiting |
+| ~30% | Good utilization |
+| ≥ 50% | Great / optimized utilization |
+
+## How to diagnose your jobs today?
+
+You can self-diagnose using these framework-agnostic methods.
+
+### Method A: Weights & Biases (Recommended)
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: Sorry I should I flagged this earlier but I'm not sure it's _the_ recommended method. It's definitely recommended as it's easy / free in terms of efforts but it cannot replace a deeper analyses that will be covered in the profiling guide

```suggestion
### Method A: Weights & Biases
```

File: docs/userguides/gpu_efficiency.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+nvidia-smi
+
+# High power draw (Watts) is usually a good signal of active GPU utilization.
+```
+
+!!! tip
+    For deeper profiling (SM occupancy, memory bandwidth, kernel-level
+    metrics), WandB system monitoring (Method A) is the recommended approach
+    on Mila nodes.
+
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: WandB will not allow follow SM occupancy nor kernel-level metrics. I think we could drop this?


```suggestion
```

File: docs/userguides/gpu_efficiency.md (hunk 9 above)
Comment (satyaog) [reply docs/userguides/gpu_efficiency.md:90]: At one point, it might be hard to track everywhere in the docs where we've given examples of high-end vs low-end GPUs. we can keep these for now but in the future we might want to give that information updated in one place (maybe https://docs.mila.quebec/technical_reference/clusters/mila/nodes/) and refer to the proper section in that page?

File: docs/userguides/gpu_efficiency.md [hunk 12]
```diff
@@ -99,10 +104,12 @@ efficient GPU utilization.
 4. **Scaling GPUs to fix I/O bottlenecks:** Do not add more GPUs if your
    pipeline is bottlenecked by storage read latency or CPU preprocessing — you
    will just idle more hardware.
-5. **Underutilizing VRAM:** Do not run with low batch sizes if your VRAM usage
-   is under 20%. Increase the batch size to maximize SM occupancy.
+5. **Underutilizing VRAM:** If your VRAM usage is under 20%, consider
+   switching to a smaller GPU or using job packing (multiple smaller jobs on
+   the same node) before increasing batch size, as larger batches may affect
+   model convergence.
```
Comment (satyaog) [inline docs/userguides/gpu_efficiency.md]: Ordered in how easy to apply.

Just as note, my previous comment on this was not to discourage users to increase batch size, but not to solely recommend increasing the batch size when VRAM usage is low. For small datasets this is simply not a solution and other solutions should be proposed to the user.

```suggestion
5. **Underutilizing VRAM:** If your VRAM usage is under 20%, consider increasing
   batch size, switching to a smaller GPU or using job packing (multiple smaller
   jobs on the same node).
```

**Review by satyaog — COMMENTED** [review]
Few small changes to be discussed I think but I would overall accept this PR

File: docs/userguides/gpu_efficiency.md (hunk 9 above)
Comment (philippepicard-stack) [reply docs/userguides/gpu_efficiency.md:90]: Indeed, it should appear in the /nodes pages, and then be used as a reference throughout the documentation

**Review by satyaog — APPROVED** [review]

File: docs/userguides/gpu_efficiency.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+## How to diagnose your jobs today?
+
+You can self-diagnose using these framework-agnostic methods.
+
+### Method A: Weights & Biases
+
+If you use WandB, go to the **System** tab of your run to find data on GPU
+utilization, CPU usage, and memory. See
+[Diagnose training bottlenecks](../userguides/wandb.md#diagnose-training-bottlenecks)
```
Comment (soline-b) [inline docs/userguides/gpu_efficiency.md]:
```suggestion
[Diagnose training bottlenecks](../../userguides/wandb.md#diagnose-training-bottlenecks)
```

**Review by soline-b — APPROVED** [review]
